Publish to Elasticsearch in deployer (#2417)
mdn/yari: deployer/README.md

and if the deployment package is different from what is already in AWS,
it will upload and publish a new version.

## Elasticsearch indexing

You just need a URL (or host name) for an Elasticsearch server and the
root of the build directory. The command will trawl all `index.json` files

> and extract all metadata and blocks of prose which gets HTML stripped.

Reviewer: ```suggestion
and extract all metadata and blocks of prose which get HTML stripped.
```

Author: I think it should probably have "its" inserted.
```suggestion
and extract all metadata and blocks of prose which gets its HTML stripped.
```
what do you think?

Reviewer: Since it's plural (`blocks`), this reads better I think:
```suggestion
and extract all metadata and blocks of prose which get their HTML stripped.
```

The command is:

```sh
cd deployer
poetry run deployer search-index --help
```

> If you have built the whole site (or partially) you simply pointed to

Reviewer: ```suggestion
If you have built the whole site (or partially) you simply point to
```

it with the first argument:

```sh
poetry run deployer search-index ../client/build
```

But by default, it does not specify the Elasticsearch URL/host. You can
either use:

```sh
export ELASTICSEARCH_URL=http://localhost:9200
poetry run deployer search-index ../client/build
```

...or...

```sh
poetry run deployer search-index ../client/build --url http://localhost:9200
```

**Note!** If you don't specify either the environment variable or the `--url`
option, the script will _not_ fail (ie. exit non-zero).

> This is to make it conveniently in GitHub Actions to control the

Reviewer: ```suggestion
This is to make it convenient in GitHub Actions to control the
```

> execution of it purely on the presence of the

Reviewer: ```suggestion
execution purely on the presence of the
```

environment variable.

### To update or not start a fresh

The default behavior is that it deletes the index first and immediately creates
it again. You can switch this off by using the `--update` option. Then it
will "cake on" the documents. So if something has been deleted since the last
build, you would still have that "stuck" in Elasticsearch.

Deleting and re-creating the index is fast so it's relatively safe to use often.

> But the indexing can take many seconds and whilst indexing, Elasticsearch

Reviewer: ```suggestion
But the indexing can take many seconds and while indexing, Elasticsearch
```

can only search what's been indexed so far.

An interesting pattern would be to use `--update` most of the time and
only from time to time omit it for a fresh new start.

But note, if you omit the `--update` (i.e. recreating the index), search

> will work. It just just find less that it finds when it's fully indexed.

Reviewer: ```suggestion
will work. It just may find less that it finds when it's fully indexed.
```

### Priority prefixes

When you index without `--update` it will delete and recreate the index.

> That means that during the time you're indexing, the searches that are happened

Reviewer: ```suggestion
That means that during the time you're indexing, the searches that are happening
```

concurrently will not find much. That's probably OK in most cases but you
can adjust the priority of what gets indexed first. This has the advantage
that most searches, that are expecting to find content in the popular document,

> will get something useful whilst the indexing is going on.

Reviewer: ```suggestion
will get something useful while the indexing is going on.
```
